### PR TITLE
Implement ComplexExpand

### PR DESCRIPTION
**Pull request by vitrun**
Treating all the variables as real numbers might be more intuitive than manipulating the expression tree.

### CODE REVIEW

**Comment by vitrun** [comment]
It seems that Sympy doesn't use Euler's formula to expand `3^(I x)`

**Comment by vitrun** [comment]
Hey @rocky , I'm submitting my first pull request to fix https://github.com/Mathics3/mathics-core/issues/788. Could you take a look at it ? Also, I'm considering rebasing it and submitting directly to the master branch. What do you think?

**Comment by rocky** [comment]
It looks like tests are failing. Run isort and black over `mathics/eval/hyperbolic.py`. I will try to look at over the weekened.

File: mathics/eval/hyperbolic.py [hunk 1]
```diff
@@ -3,8 +3,13 @@
...
 
+from sympy import Symbol
 
 def eval_ComplexExpand(expr, vars):
     sympy_expr = expr.to_sympy()
-    # Turn Re[x] -> x and remove Im[x] for all variables X that are not in vars.
+    sympy_vars = {v.to_sympy() for v in vars.elements}
+    # All vars are assumed to be real
+    replaces = [(fs, Symbol(fs.name, real=True)) for fs in sympy_expr.free_symbols 
+                if fs in sympy_vars]
```
Comment (mmatera) [inline mathics/eval/hyperbolic.py:13]: in WMA, the vars in `vars` are assumed to be complex, so the condition would be

`if fs not in sympy_vars`

File: mathics/eval/hyperbolic.py [hunk 2]
```diff
@@ -3,8 +3,13 @@
 """
 from mathics.core.convert.sympy import from_sympy
 
+from sympy import Symbol
```
Comment (mmatera) [inline mathics/eval/hyperbolic.py:6]: Maybe it would be better to use an alias `from sympy import Symbol as SympySymbol` for clarity.

**Comment by mmatera** [comment]
In https://github.com/Mathics3/mathics-core/blob/a2e926783926aa3e8902ef912faeec4077bcbe68/mathics/builtin/numbers/hyperbolic.py#L278, the docstring with the pattern is missing.

**Comment by mmatera** [comment]
With the changes I mentioned before, the tests pass in my machine.

**Comment by rocky** [comment]
@mmatera thanks for the review. 

@vitrun Although I said I'd look at over the weekend, perhaps you should first address all of the problems mentioned so far.

**Comment by vitrun** [comment]
> In
> 
> https://github.com/Mathics3/mathics-core/blob/a2e926783926aa3e8902ef912faeec4077bcbe68/mathics/builtin/numbers/hyperbolic.py#L278
> 
> , the docstring with the pattern is missing.

Hey @mmatera , thanks for the review.  Could you provide more details about what needs to be documented?  I've read through [guidelines-for-writing-documentation](https://mathics-development-guide.readthedocs.io/en/latest/extending/developing-code/extending/documentation-markup.html#guidelines-for-writing-documentation) but still not entirely clear about it.

**Comment by rocky** [comment]
I recently added https://mathics-development-guide.readthedocs.io/en/latest/extending/developer-setup.html#developer-python-code-checking-packages to mention isort and black. 

Also follow the links of the red X's above for example https://github.com/Mathics3/mathics-core/actions/runs/5311304803/jobs/9615875059?pr=869  From that you can see what specifically failed.

**Comment by mmatera** [comment]
> > In
> > https://github.com/Mathics3/mathics-core/blob/a2e926783926aa3e8902ef912faeec4077bcbe68/mathics/builtin/numbers/hyperbolic.py#L278
> > 
> > , the docstring with the pattern is missing.
> 
> Hey @mmatera , thanks for the review. Could you provide more details about what needs to be documented? I've read through [guidelines-for-writing-documentation](https://mathics-development-guide.readthedocs.io/en/latest/extending/developing-code/extending/documentation-markup.html#guidelines-for-writing-documentation) but still not entirely clear about it.

Actually, is not about the documentation, but the way in which a Builtin definition becomes into a Definition:
each `eval` method requires a python docstring with a pattern. When the interpreter finds that pattern, then execute the code. Before the line of the comment, we need a docstring ```"""ComplexExpand[expr_]"""```  to tell the interpreter that 
each time if founds an expression of the form  `ComplexExpand[something]` it must call `ComplexExpand.eval(something)`
So, the definition should look like:

```
    def eval(self, expr, evaluation: Evaluation):
        """ComplexExpand[expr_]"""
        return eval_ComplexExpand(expr, ListExpression())
```
in order to have any effect.  Notice that the named pattern `expr_` must have the same name that the parameter of the function.

Please see https://mathics-development-guide.readthedocs.io/en/latest/extending/developing-code/extending/tutorial.html, and in particular, this:https://mathics-development-guide.readthedocs.io/en/latest/extending/developing-code/extending/tutorial/1-builtin.html  for more details

**Comment by vitrun** [comment]
> With the changes I mentioned before, the tests pass in my machine.



> I recently added https://mathics-development-guide.readthedocs.io/en/latest/extending/developer-setup.html#developer-python-code-checking-packages to mention isort and black.
> 
> Also follow the links of the red X's above for example https://github.com/Mathics3/mathics-core/actions/runs/5311304803/jobs/9615875059?pr=869 From that you can see what specifically failed.

Thanks, I just did it. Now the unit test is blocking me with following failures. To debug it, I tried to set a breakpoint in `builtin/image/test.py` with no luck. I just doesn't stop there.
```
FAILED test/builtin/drawing/test_image.py::test_image[BinaryImageQ[hedy]-False-] - assert <Expression: System`BinaryImageQ[-Image-]> == <SymbolConstant: System`False>
FAILED test/builtin/drawing/test_image.py::test_image[BinaryImageQ[Binarize[hedy]]-True-] - assert <Expression: System`BinaryImageQ[-Image-]> == <SymbolConstant: System`True>
```

**Comment by vitrun** [comment]
> > With the changes I mentioned before, the tests pass in my machine.
> 
> > I recently added https://mathics-development-guide.readthedocs.io/en/latest/extending/developer-setup.html#developer-python-code-checking-packages to mention isort and black.
> > Also follow the links of the red X's above for example https://github.com/Mathics3/mathics-core/actions/runs/5311304803/jobs/9615875059?pr=869 From that you can see what specifically failed.
> 
> Thanks, I just did it. Now the unit test is blocking me with following failures. To debug it, I tried to set a breakpoint in `builtin/image/test.py` with no luck. I just doesn't stop there.
> 
> ```
> FAILED test/builtin/drawing/test_image.py::test_image[BinaryImageQ[hedy]-False-] - assert <Expression: System`BinaryImageQ[-Image-]> == <SymbolConstant: System`False>
> FAILED test/builtin/drawing/test_image.py::test_image[BinaryImageQ[Binarize[hedy]]-True-] - assert <Expression: System`BinaryImageQ[-Image-]> == <SymbolConstant: System`True>
> ```

Update:
`test_image` passed after installing `requirements-full.txt`, but `make check` still fails:
```
5569 Tests for 1089 built-in symbols, 5566 passed, 3 failed, 0 skipped.
Failed:
  - OpenRead in Reference of Built-in Symbols / File and Stream Operations
  - TextRecognize in Reference of Built-in Symbols / Image Manipulation
  - OpenRead in Reference of Built-in Symbols / Input/Output, Files, and Filesystem
```

Seems that they also have a chance of failing in the commit before my commits. I tested it in commit `a2e926`.

**Comment by mmatera** [comment]
As far as I can see, the problem is that this branch requires a rebase from master. I am going to rebase [Mathics3:add-ComplexExpand](https://github.com/Mathics3/mathics-core/tree/add-ComplexExpand), then, I hope this starts to work...

**Comment by mmatera** [comment]
@rocky, I think this ready to merge.

**Comment by mmatera** [comment]
Thanks @vitrun for the work!

**Comment by rocky** [comment]
> @rocky, I think this ready to merge.

@mmatera Ok - thanks for the review and yes @vitrun  thanks for the work.

**Comment by vitrun** [comment]
Thank you, @mmatera and @rocky ,  for your reviews and guides.

**Pull request merged**